### PR TITLE
Valgrind tutorial

### LINKED ISSUE #23: Valgrind tutorial
Explain how 
- [ ] Valgrind works  
- [ ] why it is important

### PR DESCRIPTION
**Pull request by Luthaf**
Fix #23

Some tests fail when using valgrind, which mean we can not currently run valgrind in CI. See #137, #205, #206.

- cpplint 'tests' also fail under valgrind, because they invoke `make` which leaks memory.
- python bindings test leak memory/use uninitialized variables inside Python interpreter

Also, related to #121, `test_calculators` takes 10 **minutes**!! to finish when running with valgrind.

We can still merge this, and enable valgrind in CI once everything above is fixed.

-----

- [X] Add tests, examples, and complete documentation of any added functionality
    - [X] Write additional documentation in the Sphinx (not docstrings!) to
          explain the feature and its usage in plain English

### CODE REVIEW

**Comment by felixmusil** [comment]
I don't see why we would like to run valgrind on anything but the c++ tests so the two limits that you mention should not even be I think. 
To do so wouldn't it be possible to run valgrind on  ctest with only on the c++ tests ?
Is it what `ctest -T memcheck` does already ?

Also I would not include valgrind in the regular CI workflow since it will always take too much time.

**Comment by max-veit** [comment]
>     * cpplint 'tests' also fail under valgrind, because they invoke `make` which leaks memory.

...what? Why are you running `valgrind` on `cpplint`? This sounds very much like a "you're doing it wrong".

>     * python bindings test leak memory/use uninitialized variables inside Python interpreter

Is the problem with the Python binding code itself? Otherwise I'd say it seems like the wrong tool to use for the job.

**Comment by Luthaf** [comment]
> To do so wouldn't it be possible to run valgrind on ctest with only on the c++ tests ?
> Is it what ctest -T memcheck does already ?

`ctest -T memcheck` run **ALL** tests using valgrind. Hence the issue with cpplint and the python bindings.

> ...what? Why are you running valgrind on cpplint? This sounds very much like a "you're doing it wrong".

Because cpplint adds itself to the tests: https://github.com/cosmo-epfl/librascal/blob/089fe8b8bba5dc9d0f403aa137d1b8819ccd864e/cmake/cpplint.cmake#L107

I am very much in favor of removing this, but this is a separate issue.

---

It is possible to only use valgrind on C++ test by using a different approach (manually setting a "test runner"), the main problem with this is that a cmake reconfigure is required when enabling/disabling valgrind. If this is OK with you I can change to this.

**Comment by Luthaf** [comment]
> Is the problem with the Python binding code itself? Otherwise I'd say it seems like the wrong tool to use for the job.

No, I don't think so. It looked like problems inside the Python interpreter initialization. I don't have the logs anymore, but you can easily reproduce them.

---

> Also I would not include valgrind in the regular CI workflow since it will always take too much time.

I would, for the same reason we use multiple compilers: if we don't use it, then errors will go unnoticed for a long time. See the three issues mentioned in the first message of this PR.

**Comment by felixmusil** [comment]
Python is well known to have [memory leaks](https://stackoverflow.com/questions/3982036/how-can-i-use-valgrind-with-python-c-extensions) and special care should be taken. But again I don't see why we would like to check for that on the python side.

> It is possible to only use valgrind on C++ test by using a different approach (manually setting a "test runner"), the main problem with this is that a cmake reconfigure is required when enabling/disabling valgrind. If this is OK with you I can change to this.

What are the downside of this reconfiguration ? To me it does not make sense to run valgrind on anything else than the cpp tests so if ctest does not allow for that I would not use it and go another route.

**Comment by Luthaf** [comment]
> What are the downside of this reconfiguration ?

You have to remember to do it =) But nothing major.

> But again I don't see why we would like to check for that on the python side.

We may want to check that the code is `bindings/` is not introducing errors, since it is not trivial.

**Comment by felixmusil** [comment]
> We may want to check that the code is bindings/ is not introducing errors, since it is not trivial.

Then a dedicated valgrind test for the python side should be made.

**Comment by Luthaf** [comment]
sounds good 👍

**Comment by Luthaf** [comment]
Updated to NOT use the ctest provided memcheck and setting valgrind manually as needed on C++ and Python bindings tests.

**Comment by max-veit** [comment]
Just tried this out, and running tests with valgrind is still overwhelmingly uninformative - the Python messages are not, in fact, suppressed (loads of useless complaints of the form:

<details>
<summary>output:</summary>

    ==27954== 589,856 bytes in 1 blocks are still reachable in loss record 1,166 of 1,166
    ==27954==    at 0x483877F: malloc (vg_replace_malloc.c:309)
    ==27954==    by 0x20F165: _PyObject_Malloc (obmalloc.c:1585)
    ==27954==    by 0x20F165: PyObject_Malloc (obmalloc.c:633)
    ==27954==    by 0x26782F: new_keys_object (dictobject.c:534)
    ==27954==    by 0x26782F: dictresize (dictobject.c:1141)
    ==27954==    by 0x268878: insertion_resize (dictobject.c:994)
    ==27954==    by 0x268878: PyDict_SetDefault (dictobject.c:2820)
    ==27954==    by 0x268D16: PyUnicode_InternInPlace (unicodeobject.c:15292)
    ==27954==    by 0x2AC862: r_object (marshal.c:1173)
    ==27954==    by 0x2AC735: r_object (marshal.c:1231)
    ==27954==    by 0x2ACB75: r_object (marshal.c:1402)
    ==27954==    by 0x2AD850: read_object (marshal.c:1488)
    ==27954==    by 0x2AD850: marshal_loads_impl.isra.5 (marshal.c:1800)
    ==27954==    by 0x2AD850: marshal_loads (marshal.c.h:127)
    ==27954==    by 0x26C57C: _PyMethodDef_RawFastCallKeywords (call.c:648)
    ==27954==    by 0x26C860: _PyCFunction_FastCallKeywords (call.c:734)
    ==27954==    by 0x2D82BC: call_function (ceval.c:4568)
    ==27954==    by 0x2D82BC: _PyEval_EvalFrameDefault (ceval.c:3093)

</details>
still appear, overflowing my scrollback of 5k lines, so I can't see any actual error messages). Is there a way to run _just_ the C++ tests (and, if so, can you please add it to the docs)?

In summary, I still get several test failures with `valgrind`, see below. Is this the rationale for disabling `valgrind` tests by default? (Other than the fact that they took so long, I left to do my shopping and didn't measure the total time).
```
90% tests passed, 3 tests failed out of 30

Total Test time (real) = 702.29 sec

The following tests FAILED:
         13 - test_calculator (Failed)
         20 - test_math_hyp1f1 (Failed)
         30 - python_binding_test (Failed)
Errors while running CTest

```

**Comment by Luthaf** [comment]
> Just tried this out, and running tests with valgrind is still overwhelmingly uninformative - the Python messages are not, in fact, suppressed (loads of useless complaints of the form:

I can reproduce this, I think the issue is that if Python is compiled without debug symbols, the suppression file can not work. I removed Python tests from valgrind testing for now, I'll try to find a better way to make them work.

**Comment by max-veit** [comment]
Here's the latest command line I got before my screen filled with garbage:
```
30: Test command: /usr/bin/valgrind "--error-exitcode=100" "--leak-check=full" "--show-leak-kinds=all" "--errors-for-leak-kinds=all" "--track-origins=yes" "--dsymutil=yes" "--suppressions=/home/veit/scratch/codes/librascal/tests/valgrind.suppressions" "--suppressions=/home/veit/scratch/codes/librascal/tests/valgrind-python.suppressions" "/home/veit/scratch/miniconda3/envs/rascal-clean/bin/python3" "/home/veit/scratch/codes/librascal/tests/python/python_binding_tests.py"
```

But my Python does not seem to be missing debug symbols (emphasis mine):
```
% file $(which python3.7)
/local/scratch/mveit/miniconda3/envs/rascal-clean/bin/python3.7: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, *with debug_info, not stripped*
```
It's from Conda, FWIW.

**Comment by Luthaf** [comment]
Ok, I'll try again with conda's Python to double check.

**Comment by Luthaf** [comment]
It looks like conda's Python is still missing debug symbols, so `file` might be confused here.

I get a bunch of error looking like this:

```
==22542== 18,456 bytes in 1 blocks are still reachable in loss record 218 of 220
==22542==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==22542==    by 0x210F95: ??? (in /local/scratch/fraux/local/opt/miniconda3/bin/python3.7)
==22542==    by 0x41641EF: ???
==22542==    by 0x4174747: ???
==22542==    by 0x415659F: ???
==22542==    by 0x2A9: ???
==22542==    by 0x3FF: ???
==22542==    by 0x265E5F: ??? (in /local/scratch/fraux/local/opt/miniconda3/bin/python3.7)
==22542==    by 0x43809F: ??? (in /local/scratch/fraux/local/opt/miniconda3/bin/python3.7)
==22542==    by 0x7FF: ???
==22542==    by 0x408AC27: ???
==22542==    by 0x41641EF: ???
```

---

If I compile Python from sources with `./configure --with-pydebug --with-pymalloc --with-valgrind`, and use the Python-provided suppression file, I still get multiple errors looking like

```
72 bytes in 1 blocks are possibly lost in loss record 4 of 68
==22581==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==22581==    by 0x1B775E: _PyMem_RawMalloc (obmalloc.c:99)
==22581==    by 0x1B7650: _PyMem_DebugRawAlloc (obmalloc.c:2135)
==22581==    by 0x1B768B: _PyMem_DebugRawMalloc (obmalloc.c:2168)
==22581==    by 0x1B8C2B: PyMem_RawMalloc (obmalloc.c:572)
==22581==    by 0x1B9E1B: _PyObject_Malloc (obmalloc.c:1629)
==22581==    by 0x1B7650: _PyMem_DebugRawAlloc (obmalloc.c:2135)
==22581==    by 0x1B768B: _PyMem_DebugRawMalloc (obmalloc.c:2168)
==22581==    by 0x1B77C2: _PyMem_DebugMalloc (obmalloc.c:2317)
==22581==    by 0x1B8ED5: PyObject_Malloc (obmalloc.c:685)
==22581==    by 0x183076: PyFloat_FromDouble (floatobject.c:122)
==22581==    by 0x2A55C5: fill_time (posixmodule.c:2105)
```

and 

```
Conditional jump or move depends on uninitialised value(s)
==22581==    at 0x5827411: internal_utf8_loop (loop.c:303)
==22581==    by 0x5827411: __gconv_transform_internal_utf8 (skeleton.c:609)
==22581==    by 0x58BDEF4: wcsrtombs (wcsrtombs.c:110)
==22581==    by 0x5843B20: wcstombs (wcstombs.c:34)
==22581==    by 0x293ED1: wcstombs (stdlib.h:154)
==22581==    by 0x293ED1: encode_current_locale (fileutils.c:653)
==22581==    by 0x29404F: encode_locale_ex (fileutils.c:763)
==22581==    by 0x2940CC: encode_locale (fileutils.c:773)
==22581==    by 0x294A50: _Py_EncodeLocaleRaw (fileutils.c:809)
==22581==    by 0x29520C: _Py_wreadlink (fileutils.c:1675)
==22581==    by 0x2966C2: calculate_argv0_path (getpath.c:962)
==22581==    by 0x298966: calculate_path (getpath.c:1238)
==22581==    by 0x298E8A: _PyPathConfig_Calculate (getpath.c:1353)
==22581==    by 0x26C5BE: pathconfig_calculate (pathconfig.c:317)
==22581==  Uninitialised value was created by a stack allocation
==22581==    at 0x293E2B: encode_current_locale (fileutils.c:617)
```

---

So now I am getting confused as how I could get this to work in the first place ...

**Comment by Luthaf** [comment]
I added a valgrind builder on CI, to ensure we don't introduce new memory errors. The total test time with valgrind is around 15min (build + test at 23min), which I think is acceptable. The next slowest builder is clang-9-debug, at 4min tests/15min build + test.

This is thus fully ready for review!

**Comment by Luthaf** [comment]
I did not add valgrind tests for the Python bindings, since I could not get them to work in an helpful manner (https://github.com/cosmo-epfl/librascal/pull/207#issuecomment-555038452). They can be added later if needed.

**Comment by Luthaf** [comment]
> why use gcc-7 in this case instead of the 5 and 9 that are tested ?

I went with the default gcc for Ubuntu 18.04, but you are right this should use gcc-5 or gcc-9 (I think I prefer gcc-9, any opinion?)

> why run the tests in this build ? shouldn't it be like the lint and doc CI builds that do one thing ?

Because valgrind runs on the tests

**Comment by felixmusil** [comment]
> I went with the default gcc for Ubuntu 18.04, but you are right this should use gcc-5 or gcc-9 (I think I prefer gcc-9, any opinion?)

gcc-9 is fine.

> Because valgrind runs on the tests

That is quite clear, but I get the feeling that the tests are also run without valgrind in the CI build. I might be mistaken though.

**Comment by Luthaf** [comment]
> That is quite clear, but I get the feeling that the tests are also run without valgrind in the CI build. I might be mistaken though.

Setting `RASCAL_TESTS_USE_VALGRIND=ON` changes the way the tests are run to use valgrind all the time.

**Comment by felixmusil** [comment]
> Setting RASCAL_TESTS_USE_VALGRIND=ON changes the way the tests are run to use valgrind all the time.

I see. Could you document this in the cmake file next to `option(RASCAL_TESTS_USE_VALGRIND "run tests using valgrind" OFF)` so that is very clear to all ?

**Comment by Luthaf** [comment]
What would you add in addition to `"run tests using valgrind"`? `"always run tests using valgrind"`/"run al c++ tests using valgrind"? Something else?

**Comment by felixmusil** [comment]
I would go for "Configure all c++ tests to be ran WITHIN Valgrind"

**Comment by Luthaf** [comment]
Ping @felixmusil @max-veit this is ready for review :smiley:

**Comment by Luthaf** [comment]
> Oh, fun, seems I'm getting a new error related to adaptor_filter:

Is that on this branch or on master?

File: README.rst [hunk 1]
```diff
@@ -151,9 +151,12 @@ build and run the tests:
    make
    ctest -V
 
+You can also run the tests with Valgrind (a memory-errors checker) by passing
```
[the remark this answers is not in the document]
Comment (Luthaf) [reply README.rst]: fair enough =)

**Comment by Luthaf** [comment]
>Oh, fun, seems I'm getting a new error related to adaptor_filter: 

CI is seeing it too =/ Let's open an issue !

**Comment by max-veit** [comment]
Do you want to add the error to the ignore-list so that we can get CI working before merging this branch?

**Comment by Luthaf** [comment]
Sure, how do you do that?

**Comment by max-veit** [comment]
...I thought it would be the same as ignoring the other existing errors, with a `valgrind.suppressions` file. No idea how to specify error filters there though.

**Comment by Luthaf** [comment]
Oh, right. Did not though of that

**Comment by Luthaf** [comment]
Rebased and updated, should be good to go!
